presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -277,29 +279,31 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
 
         private Multimap<VariableReferenceExpression, VariableReferenceExpression> extractCorrelatedSymbolsMapping(List<Expression> correlatedConjuncts)
         {
-            // TODO: handle coercions and non-direct column references
             ImmutableMultimap.Builder<VariableReferenceExpression, VariableReferenceExpression> mapping = ImmutableMultimap.builder();
             for (Expression conjunct : correlatedConjuncts) {
+                // TODO: Handle non-direct column references for ArithmeticBinaryExpression, ArithmeticUnaryExpression, NotExpression, etc
                 if (!(conjunct instanceof ComparisonExpression)) {
                     continue;
                 }
 
                 ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                if (!(comparison.getLeft() instanceof SymbolReference
-                        && comparison.getRight() instanceof SymbolReference
-                        && comparison.getOperator().equals(EQUAL))) {
-                    continue;
-                }
-
-                VariableReferenceExpression left = variableAllocator.toVariableReference(comparison.getLeft());
-                VariableReferenceExpression right = variableAllocator.toVariableReference(comparison.getRight());
-
-                if (correlation.contains(left) && !correlation.contains(right)) {
-                    mapping.put(left, right);
-                }
-
-                if (correlation.contains(right) && !correlation.contains(left)) {
-                    mapping.put(right, left);
+                // handle coercions and non-direct column references
+                Expression leftExpression = new ExpressionEvaluationVisitor().process(comparison.getLeft());
```

Comment:
only create the ExpressionEvaluationVisitor once and reuse.

Reply:
ok